text/0000-deprecate-property-lookup-fallback.md

```diff
@@ -0,0 +1,126 @@
+- Start Date: 2-15-2018
+- RFC PR:
+- Ember Issue:
+
+# Summary
+
+Deprecate the fallback behavior of resolving `{{foo}}` by requiring the usage of `{{this.foo}}` as syntax to refer to properties of the templates's backing component. This would be the default behavior in Glimmer Components.
+
+For example, given the following component class:
+
+```js
+import Component from '@ember/component';
+export default Component.extends({
+  init() {
+    super(...arguments);
+    this.set('greeting', 'Hello');
+  }
+});
+```
+
+One would refer to the `greeting` property as such:
+
+```hbs
+<h1>{{this.greeting}}, Chad</h1>
+```
+
+Ember will render "Hello, Chad".
+
+# Motivation
+
+Currently, the way to access properties on a components class is `{{greeting}}` from a template. This works because the component class is one of the objects we resolve against during the evaluation of the expression.
+
+The first problem with this approach is that the `{{greeting}}` syntax is ambiguous, as it could be referring to a local variable (block param), a helper with no arguments, a closed over component, or a property on the component class.
+
+## Exemplar
+Consider the following example where the ambiguity can cause issues:
+
+You have a component class that looks like the following component and template:
+
+```js
+import Component from '@ember/component';
+import computed from '@ember/computed';
+
+export default Component.extend({
+  formatName: computed('firstName', 'lastName', function() {
+    return `${this.firstName} ${this.lastName}`;
+  });
+});
+```
+
+```hbs
+<h1>Hello {{formatName}}!</h1>
+```
+
+Given `{ firstName: 'Chad', lastName: 'Hietala' }` Ember will render the following:
+
+```html
+<h1>Hello Chad Hietala!</h1>
+```
+
+Now some time goes on and someone adds a `formatName` helper that looks like the following:
+
+```js
+export default function formatName([firstName, lastName]) {
+  return `${this.firstName} ${this.lastName}`;
+}
+```
+
+Due to the fact helpers take precedence over property lookups our `{{formatName}}` now resolves to a helper. When the helper runs it doesn't have any arguments so our template now renders the following:
+
+```html
+<h1>Hello !</h1>
+```
+
+This can be a refactoring hazord and can often lead to confusion for readers of the template. Upon encountering `{{greeting}}` in a component's template, the reader has to check all of these places: first you need to scan the surrounding lines for block params with that name; next you check in the helpers folder to see it there is a helper with that name (it could also be coming from an addon!); finally, you check the component's JavaScript class to look for a (computed) property.
```

Comment:
"hazard"

Reply:
Note, this problem makes it very hard to "tool" -- "surrounding block param" is purely local, so you can write a syntax colorer or completion tool that understands it. But other kinds of fallback are much less local and are therefore much harder to tool.